File: apps/web/lib/api/cn-market.ts
/**
 * A股榜单 / 情绪 / 行业板块 API client(0023 阶段③ · 3.2)。
 *
 * GET /api/v1/cn/board · 一次取:情绪条 + 涨幅/跌幅/成交额 3 榜 + 行业板块。
 * 数据 Sina 源 · 只读。换手率/量比本期界面不体现(Sina 无字段 · 东财不可达)。
 */

import { fetchWithRetry } from './fetch-with-retry'

const API_BASE = process.env.NEXT_PUBLIC_API_URL ?? '/api-proxy'

export interface CnSpotRow {
  symbol: string // 纯代码 '600519'
  name: string
  last_price: number
  change_pct: number // %(可负)
  change_amount: number
  amount: number // 成交额(元)
  volume: number
}

export interface CnBreadth {
  ts: string
  up_count: number
  down_count: number
  flat_count: number
  limit_up_count: number // 估算
  limit_down_count: number // 估算
  total_amount: number // 两市总成交额(元)
}

export interface CnSector {
  name: string
  change_pct: number
  stock_count: number
  total_amount: number
  leader_name: string
  leader_change_pct: number
}

export interface CnBoardResponse {
  breadth: CnBreadth | null
  data_as_of: string | null
  pool_size?: number
  scope_label?: string
  gainers: CnSpotRow[]
  losers: CnSpotRow[]
  top_amount: CnSpotRow[]
  sectors: CnSector[]
}

export class CnMarketApiError extends Error {
  constructor(
    public readonly status: number,
    public readonly detail: string,
  ) {
    super(`CnMarketApi ${status}: ${detail}`)
    this.name = 'CnMarketApiError'
  }
}

async function getJson<T>(path: string, signal?: AbortSignal): Promise<T> {
  const r = await fetchWithRetry(`${API_BASE}${path}`, { signal })
  if (!r.ok) {
    let detail = `HTTP ${r.status}`
    try {
      const body = (await r.json()) as { detail?: unknown }
      if (typeof body.detail === 'string') detail = body.detail
    } catch {
      /* ignore */
    }
    throw new CnMarketApiError(r.status, detail)
  }
  return (await r.json()) as T
}

/** A股榜单 + 情绪条 + 行业板块(一次取齐)· limit 默认 100 = 各榜显示前 100。 */
export function fetchCnBoard(limit = 100, signal?: AbortSignal): Promise<CnBoardResponse> {
  return getJson<CnBoardResponse>(`/api/v1/cn/board?limit=${limit}`, signal)
}

/**
 * A股全市场搜索(代码 / 中文名 · 覆盖全 ~5500)。
 * 走后端 /cn/search(查 cn_spot_snapshot 全市场快照)· 返带价/涨跌的行,可直接渲染 + 点进详情。
 */
export function searchCnSpot(q: string, limit = 30, signal?: AbortSignal): Promise<CnSpotRow[]> {
  return getJson<CnSpotRow[]>(
    `/api/v1/cn/search?q=${encodeURIComponent(q)}&limit=${limit}`,
    signal,
  )
}

File: apps/api/app/schemas/cn_market.py
"""A股榜单 / 情绪 / 行业板块 契约(0023 阶段③ · 3.2)。

数据源:Sina(stock_zh_a_spot 全市场 + stock_sector_spot 行业板块)· 东财 _em 本地不可达。
红线:只读行情。frozen + extra=forbid · ts 全 tz-aware UTC。

本期榜单只做 3 个维度(涨幅/跌幅/成交额)· 换手率/量比 Sina 无该字段、东财不可达 →
本期不做、界面不体现(后续待办)。
"""

from __future__ import annotations

from pydantic import AwareDatetime, BaseModel, ConfigDict, Field


class CnSpotRow(BaseModel):
    """A股个股实时快照单行(Sina stock_zh_a_spot)。"""

    model_config = ConfigDict(extra="forbid", frozen=True)

    symbol: str = Field(min_length=1, description="纯代码(去 sh/sz/bj 前缀)· 如 '600519'")
    name: str = Field(min_length=1)
    last_price: float = Field(ge=0, description="最新价(停牌可能为 0)")
    change_pct: float = Field(description="涨跌幅 %(已是百分数 · 可负)")
    change_amount: float = Field(description="涨跌额(可负)")
    amount: float = Field(ge=0, description="成交额(元)")
    volume: float = Field(ge=0, description="成交量(股)")


class CnBreadth(BaseModel):
    """A股市场情绪条 · 全市场快照精确聚合 + 涨跌停估算。"""

    model_config = ConfigDict(extra="forbid", frozen=True)

    ts: AwareDatetime
    up_count: int = Field(ge=0, description="上涨家数(涨跌幅 > 0)")
    down_count: int = Field(ge=0, description="下跌家数(涨跌幅 < 0)")
    flat_count: int = Field(ge=0, description="平盘家数(涨跌幅 = 0 · 含停牌)")
    limit_up_count: int = Field(
        ge=0, description="涨停家数(估算 · 按板块涨跌幅阈值 · 东财涨跌停池不可达)",
    )
    limit_down_count: int = Field(ge=0, description="跌停家数(估算 · 同上口径)")
    total_amount: float = Field(ge=0, description="两市总成交额(元)")


class CnSector(BaseModel):
    """A股行业板块单行(Sina stock_sector_spot · 新浪行业)。"""

    model_config = ConfigDict(extra="forbid", frozen=True)

    name: str = Field(min_length=1, description="板块名(如 '玻璃行业')")
    change_pct: float = Field(description="板块涨跌幅 %(可负)")
    stock_count: int = Field(ge=0, description="成分股家数")
    total_amount: float = Field(ge=0, description="板块总成交额(元)")
    leader_name: str = Field(default="", description="代表/领涨股名")
    leader_change_pct: float = Field(default=0.0, description="代表股涨跌幅 %")


class CnBoardResponse(BaseModel):
    """`GET /api/v1/cn/board` 响应 · 情绪条 + 3 榜单 + 行业板块。

    榜单是同一份全市场快照的 3 种排序(涨幅/跌幅/成交额)· 前端 Tab 内存切换。
    """

    model_config = ConfigDict(extra="forbid", frozen=True)

    breadth: CnBreadth | None = Field(default=None, description="情绪条 · 无数据为 null")
    data_as_of: AwareDatetime | None = Field(default=None, description="快照最新时间 · 截至")
    gainers: list[CnSpotRow] = Field(description="涨幅榜(change_pct DESC)")
    losers: list[CnSpotRow] = Field(description="跌幅榜(change_pct ASC)")
    top_amount: list[CnSpotRow] = Field(description="成交额榜(amount DESC)")
    sectors: list[CnSector] = Field(description="行业板块(change_pct DESC)")
